File: scripts/update_secret.py
"""
GitHub Secret 更新模块
使用 GitHub API + PyNaCl 加密更新 Repository Secret
"""

import base64
import requests
from nacl import encoding, public


def update_secret(token: str, owner: str, repo: str, secret_name: str, secret_value: str):
    """更新 GitHub Repository Secret"""
    headers = {
        'Authorization': f'Bearer {token}',
        'Accept': 'application/vnd.github+json',
        'X-GitHub-Api-Version': '2022-11-28',
    }
    
    # 1. 获取仓库公钥
    key_url = f'https://api.github.com/repos/{owner}/{repo}/actions/secrets/public-key'
    key_response = requests.get(key_url, headers=headers, timeout=30)
    key_response.raise_for_status()
    key_data = key_response.json()
    
    public_key = key_data['key']
    key_id = key_data['key_id']
    
    # 2. 使用 PyNaCl 加密
    encrypted_value = encrypt_secret(public_key, secret_value)
    
    # 3. 更新 Secret
    update_url = f'https://api.github.com/repos/{owner}/{repo}/actions/secrets/{secret_name}'
    update_response = requests.put(
        update_url,
        headers=headers,
        json={
            'encrypted_value': encrypted_value,
            'key_id': key_id,
        },
        timeout=30,
    )
    update_response.raise_for_status()


def encrypt_secret(public_key_b64: str, secret_value: str) -> str:
    """使用 libsodium sealed box 加密"""
    public_key_bytes = base64.b64decode(public_key_b64)
    sealed_box = public.SealedBox(public.PublicKey(public_key_bytes))
    encrypted = sealed_box.encrypt(secret_value.encode('utf-8'))
    return base64.b64encode(encrypted).decode('utf-8')

File: scripts/update-secret.js
/**
 * GitHub Secret 更新模块
 * 使用 GitHub API + libsodium 加密更新 Repository Secret
 */

import sodium from 'libsodium-wrappers';

/**
 * 更新 GitHub Repository Secret
 * @param {string} token - GitHub Personal Access Token
 * @param {string} owner - 仓库所有者
 * @param {string} repo - 仓库名称
 * @param {string} secretName - Secret 名称
 * @param {string} secretValue - Secret 值
 */
export async function updateSecret(token, owner, repo, secretName, secretValue) {
  await sodium.ready;

  // 1. 获取仓库公钥
  const keyResponse = await fetch(
    `https://api.github.com/repos/${owner}/${repo}/actions/secrets/public-key`,
    {
      headers: {
        'Authorization': `Bearer ${token}`,
        'Accept': 'application/vnd.github+json',
        'X-GitHub-Api-Version': '2022-11-28',
      },
    }
  );

  if (!keyResponse.ok) {
    throw new Error(`获取公钥失败: ${keyResponse.status} ${await keyResponse.text()}`);
  }

  const { key, key_id } = await keyResponse.json();

  // 2. 使用 libsodium 加密 secret
  const publicKey = sodium.from_base64(key, sodium.base64_variants.ORIGINAL);
  const messageBytes = sodium.from_string(secretValue);
  const encryptedBytes = sodium.crypto_box_seal(messageBytes, publicKey);
  const encryptedValue = sodium.to_base64(encryptedBytes, sodium.base64_variants.ORIGINAL);

  // 3. 更新 secret
  const updateResponse = await fetch(
    `https://api.github.com/repos/${owner}/${repo}/actions/secrets/${secretName}`,
    {
      method: 'PUT',
      headers: {
        'Authorization': `Bearer ${token}`,
        'Accept': 'application/vnd.github+json',
        'X-GitHub-Api-Version': '2022-11-28',
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        encrypted_value: encryptedValue,
        key_id: key_id,
      }),
    }
  );

  if (!updateResponse.ok) {
    throw new Error(`更新 Secret 失败: ${updateResponse.status} ${await updateResponse.text()}`);
  }
}
